File: ecobertura.core/src/ecobertura/core/CorePlugin.java
package ecobertura.core;

import java.util.logging.Logger;

import org.eclipse.core.runtime.Plugin;
import org.osgi.framework.BundleContext;

import ecobertura.core.log.EclipseLogger;
import ecobertura.core.state.PluginState;
import ecobertura.core.trace.Trace;

/**
 * Controls the plug-in life cycle of the eCobertura Core.
 */
public class CorePlugin extends Plugin {

	private static final Logger logger = Logger.getLogger(CorePlugin.PLUGIN_ID);
	public static final String PLUGIN_ID = "ecobertura.core"; //$NON-NLS-1$

	private static CorePlugin instance;
	
	private PluginState pluginState;
	
	@Override
	public void start(BundleContext context) throws Exception {
		super.start(context);
		instance = this;
		
		Trace.configureForPluginId(PLUGIN_ID);
		EclipseLogger.logFor(getLog());

		pluginState = PluginState.initialize(getStateLocation());
		
		logger.info("plugin started"); //$NON-NLS-1$
	}

	@Override
	public void stop(BundleContext context) throws Exception {
		pluginState.cleanUp();
		instance = null;
		super.stop(context);
		logger.info("plugin stopped"); //$NON-NLS-1$
	}
	
	public static CorePlugin instance() {
		return instance;
	}
	
	public PluginState pluginState() {
		return pluginState;
	}
}

File: ecobertura.core/src/ecobertura/core/CorePlugin.scala
package ecobertura.core

import java.util.logging.Logger

import org.eclipse.core.runtime.Plugin
import org.osgi.framework.BundleContext

import ecobertura.core.log.EclipseLogger
import ecobertura.core.results.CoverageResultsCollector
import ecobertura.core.state.PluginState
import ecobertura.core.trace.Trace

object CorePlugin {
	private val internalPluginId = "ecobertura.core" //$NON-NLS-1$
	private val logger = Logger getLogger internalPluginId
	
	private var internalInstance: CorePlugin = null
	
	def instance = internalInstance
	def pluginId = internalPluginId
}

class CorePlugin extends Plugin {
	import CorePlugin._
	
	private var internalPluginState: PluginState = null
	private var internalResultsCollector: CoverageResultsCollector = null
	
	def pluginState = internalPluginState
	def coverageResultsCollector = internalResultsCollector

	override def start(context: BundleContext): Unit = {
		super.start(context)
		internalInstance = this
		
		Trace configureForPluginId pluginId;
		EclipseLogger logFor getLog

		internalPluginState = PluginState initialize getStateLocation
		internalResultsCollector = CoverageResultsCollector collect
		
		logger info "plugin started" //$NON-NLS-1$
	}
	
	override def stop(context: BundleContext): Unit = {
		internalPluginState.cleanUp
		internalInstance = null
		super.stop(context)
		
		logger info "plugin stopped" //$NON-NLS-1$
	}
}
